federation-2/router-bridge/js-src/plan.ts

```diff
@@ -21,63 +47,463 @@ export class BridgeQueryPlanner {
+// function; it would be better to use that instead, though whenever we change
+// the signature algorithm it does make every operation appear to change in
+// Studio.
+//
+// In a GraphQL AST (which notably does not contain comments), the only place
+// where meaningful whitespace (or double quotes) can exist is in StringNodes.
+// So to print with reduced whitespace, we:
+// - temporarily sanitize strings by replacing their contents with hex
+// - use the default GraphQL printer
+// - minimize the whitespace with a simple regexp replacement
+// - convert strings back to their actual value We normalize all strings to
+//   non-block strings for simplicity.
+// (Unlike stripIgnoredCharacters, this does not remove commas.)
+function printWithReducedWhitespace(ast: DocumentNode): string {
+  const sanitizedAST = visit(ast, {
+    StringValue(node: StringValueNode): StringValueNode {
+      return {
+        ...node,
+        value: Buffer.from(node.value, "utf8").toString("hex"),
+        block: false,
+      };
+    },
+  });
+  const withWhitespace = print(sanitizedAST);
+  const minimizedButStillHex = withWhitespace
+    .replace(/\s+/g, " ")
+    .replace(/([^_a-zA-Z0-9]) /g, (_, c) => c)
+    .replace(/ ([^_a-zA-Z0-9])/g, (_, c) => c);
+  return minimizedButStillHex.replace(/"([a-f0-9]+)"/g, (_, hex) =>
+    JSON.stringify(Buffer.from(hex, "hex").toString("utf8"))
+  );
+}
+
+// sortAST sorts most multi-child nodes alphabetically. Using this as part of
+// your signature calculation function may make it easier to tell the difference
+// between queries that are similar to each other, and if for some reason your
+// GraphQL client generates query strings with elements in nondeterministic
+// order, it can make sure the queries are treated as identical.
+function sortAST(ast: DocumentNode): DocumentNode {
+  return visit(ast, {
+    Document(node: DocumentNode) {
+      return {
+        ...node,
+        // Use sortBy instead of sorted because 'definitions' is not optional.
+        // The sort on "kind" places fragments before operations within the document
+        definitions: sortBy(node.definitions, "kind", "name.value"),
+      };
+    },
+    OperationDefinition(
+      node: OperationDefinitionNode
+    ): OperationDefinitionNode {
+      return {
+        ...node,
+        variableDefinitions: sorted(
+          node.variableDefinitions,
+          "variable.name.value"
+        ),
+      };
+    },
+    SelectionSet(node: SelectionSetNode): SelectionSetNode {
+      return {
+        ...node,
+        // Define an ordering for field names in a SelectionSet.  Field first,
+        // then FragmentSpread, then InlineFragment.  By a lovely coincidence,
+        // the order we want them to appear in is alphabetical by node.kind.
+        // Use sortBy instead of sorted because 'selections' is not optional.
+        selections: sortBy(node.selections, "kind", "name.value"),
```

Comment:
should the alias be included in sorting keys?

Reply:
good question, possibly worth a discussion with the federation team ?